tasks/human_eval_008.rs

```diff
@@ -9,7 +9,67 @@ use vstd::prelude::*;
 
 verus! {
 
-// TODO: Put your solution (the specification, implementation, and proof) to the task here
+/// Specification for what it means to sum a sequence of numbers
+pub open spec fn sum(numbers: Seq<u32>) -> int {
+    numbers.fold_left(0, |acc: int, x| acc + x)
+}
+
+/// Specification for taking the product of a sequence of numbers
+pub open spec fn product(numbers: Seq<u32>) -> int {
+    numbers.fold_left(1, |acc: int, x| acc * x)
+}
+
+/// Show that the sum won't grow too large
+proof fn sum_bound(numbers: Seq<u32>)
+    ensures
+        sum(numbers) <= numbers.len() * u32::MAX,
+    decreases numbers.len(),
+{
+    if numbers.len() == 0 {
+    } else {
+        sum_bound(numbers.drop_last());
+    }
+}
+
+/// Implementation.  We leave the consequences of an intermediate
+/// overflow during the product calculation underspecified.
+fn sum_product(numbers: Vec<u32>) -> (result: (u64, Option<u32>))
+    requires
+        numbers.len() < u32::MAX,
+    ensures
+        result.0 == sum(numbers@),
+        result.1 matches Some(v) ==> v == product(numbers@),
+{
+    let mut sum_value: u64 = 0;
+    let mut prod_value: Option<u32> = Some(1);
+    let mut index = 0;
+    while index < numbers.len()
+        invariant
+            0 <= index <= numbers.len(),
+            numbers.len() < u32::MAX,
+            sum_value == sum(numbers@.take(index as int)),
+            prod_value matches Some(v) ==> v == product(numbers@.take(index as int)),
+            index <= numbers.len(),
+            index >= 0,
+            sum_value <= u64::MAX,
```

Comment:
I don't think this invariant is needed.

Reply:
Good point; that's left over from a previous attempt.